Delegate bounding box transformation to proj on >= 8.2

nyalldawson opened the pull request:
Fixes #59821

nyalldawson committed:
Attempt to delegate bounding box transformation to proj on >= 8.2

Refs #59821

nyalldawson committed:
Explain magic number

nyalldawson committed:
Add render masks

nyalldawson committed:
Add render mask

nyalldawson committed:
Add some tolerance to test

nyalldawson committed:
Update server reference files to proj >= 8.2 versions

nyalldawson committed:
Adapt tests for proj 8.2 results

nyalldawson committed:
Adapt tests for proj 8.2 results

nyalldawson committed:
Demote crs to 2d only crs before calling proj_trans_bounds

nyalldawson commented:
@rouault is https://github.com/qgis/QGIS/pull/60331/commits/8edbe0d422b6eea897dd4322ac4319e807a311c1 a known requirement for proj_trans_bounds?

rouault commented:
> is [8edbe0d](https://github.com/qgis/QGIS/commit/8edbe0d422b6eea897dd4322ac4319e807a311c1) a known requirement for proj_trans_bounds?

on released versions of PROJ, yes, since I fixed 3D support only in PROJ master per https://github.com/OSGeo/PROJ/pull/4333 a couple months ago

nyalldawson commented:
Thanks @rouault !

I noticed that fix is similar to my workaround in https://github.com/qgis/QGIS/commit/8edbe0d422b6eea897dd4322ac4319e807a311c1. Do you think we should:

1. Defer this whole change till newer proj is available.
2. Use the (buggy) old qgis logic when a compound crs is in play
3. Or conditionally apply the same logic as the proj fix?

nyalldawson committed:
Use compound crs workaround only on proj < 9.6

Upstream bug was fixed in https://github.com/OSGeo/PROJ/pull/4333

nyalldawson committed:
Update test to newer infrastructure

nyalldawson committed:
Fix testing for vertical axis of bound crs

nyalldawson committed:
Don't transform 2d bounding boxes when geocentric crs involved

This is meaningless and will give a misleading/useless result,
so raise a descriptive QgsCsException instead.

nyalldawson committed:
Test mask

nyalldawson commented:
Ok, this is ready for review.

I'm also interested in thoughts on backporting -- #59821 is very serious, but these changes ARE quite risky...

nyalldawson committed:
Test mask update

rouault commented:
> I'm also interested in thoughts on backporting -- #59821 is very serious, but these changes ARE quite risky...

I would go for the backport
